main/php_syslog.c

```diff
@@ -0,0 +1,74 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2017 The PHP Group                                |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Philip Prindeville <philipp@redfish-solutions.com>           |
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <stdlib.h>
+#include "php.h"
+#include "php_syslog.h"
+
+#include "zend_smart_string.h"
+
+#ifdef HAVE_STD_SYSLOG
+#define syslog std_syslog
+#endif
+
+PHPAPI void php_syslog(int priority, const char *message) /* {{{ */
+{
+	const char *ptr;
+	unsigned char c;
+	smart_string buf = {0};
+
+	for (ptr = message; ; ++ptr) {
+		c = *ptr;
+		if (c == '\0') {
+			syslog(priority, "%.*s", (int)buf.len, buf.c);
+			break;
+		}
+
+		if (((' ' <= c) && (c <= '~')) || (PG(syslog_highbit) && (c >= 0x80)))
```

Comment:
> the vulnerability linked was about escape sequences, so c< 32 is the only condition for that. Or do i miss yet more vulnerabilities?

@weltling There are syslog implementations out there which don't implement RFC-5424, and others still which implement it but disable it by default, falling back on RFC-3164, which states:

```
4.1.3 MSG Part of a syslog Packet

   The MSG part will fill the remainder of the syslog packet.  This will
   usually contain some additional information of the process that
   generated the message, and then the text of the message.  There is no
   ending delimiter to this part.  The MSG part of the syslog packet
   MUST contain visible (printing) characters.  The code set
   traditionally and most often used has also been seven-bit ASCII in an
   eight-bit field like that used in the PRI and HEADER parts.  In this
   code set, the only allowable characters are the ABNF VCHAR values
   (%d33-126) and spaces (SP value %d32).  However, no indication of the
   code set used within the MSG is required, nor is it expected.  Other
   code sets MAY be used as long as the characters used in the MSG are
   exclusively visible characters and spaces similar to those described
   above.  The selection of a code set used in the MSG part SHOULD be
   made with thoughts of the intended receiver.  A message containing
   characters in a code set that cannot be viewed or understood by a
   recipient will yield no information of value to an operator or
   administrator looking at it.
```

The only actionable part of that paragraph is, "In this code set, the only allowable characters are the ABNF VCHAR values (%d33-126) and spaces (SP value %d32)."

The rest is unactionable.  For example, "However, no indication of the code set used within the MSG is required, nor is it expected.  Other code sets MAY be used as long as the characters used in the MSG are exclusively visible characters and spaces similar to those described above."

If you don't know what the code set is because it's not explicitly called out, then how do you know if a character is "printing" or not in that code set?  You don't.  You can't.